core/frontend/src/render/webgl/glsl/Instancing.ts

```diff
@@ -20,16 +22,70 @@ float extractInstanceBit(uint flag) { return extractNthBit(a_instanceOverrides.r
 `;
 
 const computeInstancedModelMatrixRTC = `
-  g_modelMatrixRTC = mat4(
-    a_instanceMatrixRow0.x, a_instanceMatrixRow1.x, a_instanceMatrixRow2.x, 0.0,
-    a_instanceMatrixRow0.y, a_instanceMatrixRow1.y, a_instanceMatrixRow2.y, 0.0,
-    a_instanceMatrixRow0.z, a_instanceMatrixRow1.z, a_instanceMatrixRow2.z, 0.0,
-    a_instanceMatrixRow0.w, a_instanceMatrixRow1.w, a_instanceMatrixRow2.w, 1.0);
+  if (g_isAreaPattern) {
+    vec2 spacing = u_patternParams.yz;
+    float scale = u_patternParams.w;
+
+    float x = u_patternOrigin.x + a_patternX * spacing.x;
+    float y = u_patternOrigin.y + a_patternY * spacing.y;
+    vec4 translation = vec4(x / scale, y / scale, 0.0, 1.0);
+    mat4 symbolTrans = u_patOrg;
+    symbolTrans[3] = symbolTrans * translation;
+
+    g_modelMatrixRTC = u_patLocalToModel * symbolTrans * u_patSymbolToLocal;

```

Comment:
@bbastings I'd like to (1) reduce the amount of computation happening in this shader and (2) reduce precision issues introduced when the pattern symbol is defined far away from the origin. I'm having trouble getting that right - do you have an obvious suggestion?

Reply:
You have test cases where the pattern symbol isn't defined near the origin, that's super strange and unexpected?

Reply:
I specifically made one because we had issues with precision for geometry parts defined far from origin (basically connectors would define parts in world coordinates for some reason).
If you think pattern symbols are not likely to suffer similar problems, I won't worry about it. What I have already prevents precision issues when the patterned element is far from the origin.

Reply:
I don't think you should worry about it...creating parts far away from the origin (or in world coords) is idiotic...maybe the "problematic" geometry stream sniffer could also check for this?